File: leetcode/2001-2999/2558-take-gifts-from-the-richest-pile/TakeGiftsFromTheRichest.ts
// Brute force
function pickGifts(gifts: number[], k: number): number {
  while (k-- > 0) {
    const { index, value } = gifts.reduce((previousAccumulated, currentElement, currentIndex) => {
      const { value } = previousAccumulated;

      if (currentElement > value) {
        return {
          index: currentIndex,
          value: currentElement
        };
      }

      return previousAccumulated;

    }, { index: 0, value: gifts[0] });

    gifts[index] = Math.floor(Math.sqrt(value));
  }

  return gifts.reduce((previousAccumulatedValue, currentElement) => previousAccumulatedValue + currentElement, 0);
};

const gifts = [25, 64, 9, 4, 100];
const k = 4;

console.log(pickGifts(gifts, k));

File: leetcode/2001-2999/2558-take-gifts-from-the-richest-pile/TakeGiftsFromTheRichest.java
import java.util.PriorityQueue;

public class TakeGiftsFromTheRichest {
  public static void main(String[] args) {
    int[] gifts = { 25, 64, 9, 100, 4 };
    int k = 4;

    System.out.println((new Solution()).pickGifts(gifts, k));
  }
}

class Solution {
  public long pickGifts(int[] gifts, int k) {
    PriorityQueue<Integer> queue = new PriorityQueue<>((a, b) -> b - a);

    for (int gift : gifts) {
      queue.offer(gift);
    }

    while (k-- > 0) {
      int max = queue.poll();

      queue.offer((int) Math.sqrt(max));
    }

    System.out.println(queue);

    return queue.stream().mapToLong(Long::valueOf).reduce(0, Long::sum);
  }
}
